<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>Running DSC</title>
  <meta name="description" content="Fine-mapping result post processing">

  <link rel="canonical" href="https://gaow.github.io/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler.html">
  <link rel="alternate" type="application/rss+xml" title="Gao's recipe" href="https://gaow.github.io/jupyterbook-sos-demo/feed.xml">

  <meta property="og:url"         content="https://gaow.github.io/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="Running DSC" />
<meta property="og:description" content="Fine-mapping result post processing" />
<meta property="og:image"       content="" />


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage":
    "https://gaow.github.io/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler.html",
  "headline":
    "Running DSC",
  "datePublished":
    "2019-02-01T09:55:11-06:00",
  "dateModified":
    "2019-02-01T09:55:11-06:00",
  "description":
    "Fine-mapping result post processing",
  "author": {
    "@type": "Person",
    "name": "Gao Wang and others, Stephens Lab, The University of Chicago"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "https://gaow.github.io/jupyterbook-sos-demo",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "https://gaow.github.io/jupyterbook-sos-demo",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/jupyterbook-sos-demo/assets/css/styles.css">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css ">
  <link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
  <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
  <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
  <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
  <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
  <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
  <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
  <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/jupyterbook-sos-demo/images/favicon.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    CommonHTML: {
        linebreaks: {
            automatic: true,
        },
    },
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_CHTML' async></script>

  <!-- DOM updating function -->
  <script>
const runWhenDOMLoaded = cb => {
  if (document.readyState != 'loading') {
    cb()
  } else if (document.addEventListener) {
    document.addEventListener('DOMContentLoaded', cb)
  } else {
    document.attachEvent('onreadystatechange', function() {
      if (document.readyState == 'complete') cb()
    })
  }
}

// Helper function to init things quickly
initFunction = function(myfunc) {
  runWhenDOMLoaded(myfunc);
  document.addEventListener('turbolinks:load', myfunc);
};
</script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/jupyterbook-sos-demo';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="/jupyterbook-sos-demo/assets/js/anchor.min.js"  type="text/javascript"></script>
  <script>

initFunction(function () {
    anchors.add("main h1, main h2, main h3, main h4")
});

</script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="/jupyterbook-sos-demo/assets/js/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  <script src="https://unpkg.com/nbinteract-core" async></script>

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->


  <!-- Google analytics -->
  <script src="/jupyterbook-sos-demo/assets/js/ga.js" async></script>

  <!-- Clipboard copy button -->
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js" async></script>

  <!-- Load JS that depends on site variables -->
  <script>
/**
 * Set up copy/paste for code blocks
 */
const codeCellId = index => `codecell${index}`

const clipboardButton = id =>
  `<a class="btn copybtn o-tooltip--left" data-tooltip="Copy" data-clipboard-target="#${id}">
    <img src="/jupyterbook-sos-demo/assets/images/copy-button.svg" alt="Copy to clipboard">
  </a>`

// Clears selected text since ClipboardJS will select the text when copying
const clearSelection = () => {
  if (window.getSelection) {
    window.getSelection().removeAllRanges()
  } else if (document.selection) {
    document.selection.empty()
  }
}

// Changes tooltip text for two seconds, then changes it back
const temporarilyChangeTooltip = (el, newText) => {
  const oldText = el.getAttribute('data-tooltip')
  el.setAttribute('data-tooltip', newText)
  setTimeout(() => el.setAttribute('data-tooltip', oldText), 2000)
}

const addCopyButtonToCodeCells = () => {
  // If ClipboardJS hasn't loaded, wait a bit and try again. This
  // happens because we load ClipboardJS asynchronously.
  if (window.ClipboardJS === undefined) {
    setTimeout(addCopyButtonToCodeCells, 250)
    return
  }

  const codeCells = document.querySelectorAll('.input_area pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.querySelector(`pre#${id} + a`) == null) {
      codeCell.insertAdjacentHTML('afterend', clipboardButton(id));
    }
  })

  const clipboard = new ClipboardJS('.copybtn')
  clipboard.on('success', event => {
    clearSelection()
    temporarilyChangeTooltip(event.trigger, 'Copied!')
  })

  clipboard.on('error', event => {
    temporarilyChangeTooltip(event.trigger, 'Failed to copy')
  })

  // Get rid of clipboard before the next page visit to avoid memory leak
  document.addEventListener('turbolinks:before-visit', () =>
    clipboard.destroy()
  )
}

initFunction(addCopyButtonToCodeCells);
</script>

  <!-- Hide cell code -->
  
<script>
/**
Add buttons to hide code cells
*/


var setCodeCellVisibility = function(inputField, kind) {
    // Update the image and class for hidden
    var id = inputField.getAttribute('data-id');
    var codeCell = document.querySelector(`#${id}`);

    if (kind === "visible") {
        codeCell.classList.remove('hidden');
        inputField.checked = true;
    } else {
        codeCell.classList.add('hidden');
        inputField.checked = false;
    }
}

var toggleCodeCellVisibility = function (event) {
    // The label is clicked, and now we decide what to do based on the input field's clicked status
    if (event.target.tagName === "LABEL") {
        var inputField = event.target.previousElementSibling;
    } else {
        // It is the span inside the target
        var inputField = event.target.parentElement.previousElementSibling;
    }

    if (inputField.checked === true) {
        setCodeCellVisibility(inputField, "visible");
    } else {
        setCodeCellVisibility(inputField, "hidden");
    }
}


// Button constructor
const hideCodeButton = id => `<input class="hidebtn" type="checkbox" id="hidebtn${id}" data-id="${id}"><label title="Toggle cell" for="hidebtn${id}" class="plusminus"><span class="pm_h"></span><span class="pm_v"></span></label>`

var addHideButton = function () {
  // If a hide button is already added, don't add another
  if (document.querySelector('div.hidecode input') !== null) {
      return;
  }

  // Find the input cells and add a hide button
  document.querySelectorAll('div.input_area div.highlight').forEach(function (item, index) {
    if (!item.parentElement.classList.contains("hidecode")) {
        // Skip the cell if it doesn't have a hidecode class
        return;
    }

    const id = codeCellId(index)
    item.setAttribute('id', id);
    item.insertAdjacentHTML('afterend', hideCodeButton(id))

    // Set up the visibility toggle
    hideLink = document.querySelector(`#${id} + input + label`);
    hideLink.addEventListener('click', toggleCodeCellVisibility)
  });
}


// Initialize the hide buttos
var initHiddenCells = function () {
    // Add hide buttons to the cells
    addHideButton();

    // Toggle the code cells that should be hidden
    document.querySelectorAll('div.hidecode input').forEach(function (item) {
        setCodeCellVisibility(item, 'hidden');
        item.checked = true;
    })
}

initFunction(initHiddenCells);

</script>


  <!-- Load custom website scripts -->
  <script src="/jupyterbook-sos-demo/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/jupyterbook-sos-demo/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/jupyterbook-sos-demo/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  

  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="/jupyterbook-sos-demo/assets/js/lunr/lunr.min.js" type="text/javascript"></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>
</head>

  <body>
    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  <a href="https://github.com/gaow/jupyterbook-sos-demo"><img src="/jupyterbook-sos-demo/images/logo.png" class="textbook_logo" id="sidebar-logo" data-turbolinks-permanent/></a>
  <h2 class="c-sidebar__title">Gao's recipe</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li class="c-sidebar__chapter"><a class="c-sidebar__entry" href="/jupyterbook-sos-demo/search.html"><img src="/dsc-wiki/images/search.svg" alt="Search This Wiki"></a></li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/jupyterbook-sos-demo/README.html"
        >
          
          Home
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Workflows</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/jupyterbook-sos-demo/finemapping/README.html"
        >
          
          Genetic fine-mapping
        </a>

        

          
          
          
          

          

          <ul class="c-sidebar__sections ">
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry "
                  href="/jupyterbook-sos-demo/finemapping/summary_statistics_wrangler.html"
                >
                  
                  Terminology
                </a>

                
                

              </li>
              
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry "
                  href="/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping.html"
                >
                  
                  DSC configuration syntax
                </a>

                
                

              </li>
              
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry c-sidebar__entry--active"
                  href="/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler.html"
                >
                  
                  Running DSC
                </a>

                
                

              </li>
              
            
          </ul>
        
      </li>

      
    
  </ul>
  <p class="sidebar_footer">Copyright (c) 2019 <br> <a href="http://home.uchicago.edu/gaow">Gao Wang</a><br></p>
</nav>

      
      <!-- Shamelessly copied from minimal mistakes -->


<!-- TOC will only show up if it has at least one item -->


  <aside class="sidebar__right">
    <nav class="onthispage">
      <header><h4 class="nav__title"><i class="fa fa-list"></i>   Contents</h4></header>
      <ul class="toc__menu">
  <li><a href="#input-data">Input data</a></li>
  <li><a href="#output-data">Output data</a></li>
  <li><a href="#additional-data-processing">Additional data processing</a></li>
  <li><a href="#the-workflow">The workflow</a></li>
  <li><a href="#fine-mapping-results-consolidation">Fine-mapping results consolidation</a></li>
  <li><a href="#update-variant-id-via-genomic-coordinantes">Update variant ID via genomic coordinantes</a></li>
  <li><a href="#update-genomic-coordinates-via-variant-id">Update genomic coordinates via variant ID</a></li>
  <li><a href="#example-run">Example run</a></li>
</ul>
    </nav>
  </aside>


      
      <main class="c-textbook__page" tabindex="-1">
          <div class="o-wrapper">
            <div class="c-sidebar-toggle">
  <!-- We show the sidebar by default so we use .is-active -->
  <button
    id="js-sidebar-toggle"
    class="hamburger hamburger--arrowalt is-active"
  >
    <span class="hamburger-box">
      <span class="hamburger-inner"></span>
    </span>
    <span class="c-sidebar-toggle__label">Toggle Sidebar</span>
  </button>
</div>

            
<div class="buttons">




</div>


            <div class="c-textbook__content">
              <h1 id="fine-mapping-result-post-processing">Fine-mapping result post processing</h1>

<p>This pipeline consolidates results from various fine-mapping tools to uniform format, add rsID as necessary, and perform a simple “liftover” via rsID (not the formal UCSC <code class="highlighter-rouge">liftOver</code>) to generate output in HG37 and HG38 builds.</p>

<p>This pipeline was devised by Gao Wang (UChicago), with core implemention done by Kushal Dey (Harvard).</p>

<h2 id="input-data">Input data</h2>

<p>Input are results of fine-mapping pipeline <code class="highlighter-rouge">summary_statistics_finemapping.ipynb</code> in R’s <code class="highlighter-rouge">RDS</code> format for SuSiE and CAVIAR, and <code class="highlighter-rouge">pkl</code> format for DAP.</p>

<h2 id="output-data">Output data</h2>

<p>columns are:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>chr pos ref alt variant_id locus_id ...
</code></pre></div></div>

<p>where <code class="highlighter-rouge">snp_id</code> will be rsID if some annotation files on rsID are provided and the <code class="highlighter-rouge">update_variant_id</code> workflow below was executed. Otherwise it will take the default format of <code class="highlighter-rouge">chr:pos:ref:alt</code>.</p>

<p>Because each fine-mapping software can have different output, the columns after the first 6 are arbitary. For SuSiE for example they are</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>chr pos ref alt variant_id locus_id pip cs cs_size cs_purity
</code></pre></div></div>

<p>For CAVIAR they are</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>chr pos ref alt variant_id locus_id pip cs_size
</code></pre></div></div>

<p>Naturally per locus <code class="highlighter-rouge">cs</code> related information will be redundant. This is one draw-back to using a flat table output format.</p>

<h2 id="additional-data-processing">Additional data processing</h2>

<p>Here we also provide additional routines to process the data,</p>

<ol>
  <li>Update the <code class="highlighter-rouge">variant_id</code> column using external annotations, for example by rsID.</li>
  <li>“liftover” to other builds – we only support it via rsID matching</li>
</ol>

<p>Notice that only the first 5 columns are necessary for these additional operations. The columns after the fifth can be arbitary and will be kept during the process.</p>

<ul>
  <li>To trigger optional step 1, parameter <code class="highlighter-rouge">--id-map-prefix</code> and <code class="highlighter-rouge">--id-map-suffix</code> have to be valid.</li>
  <li>To trigger optional step 2, parameter <code class="highlighter-rouge">--coordinate-map-prefix</code> and <code class="highlighter-rouge">--coordinate-map-suffix</code> have to be valid.</li>
</ul>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%cd ~/GIT/github/fine-mapping
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/scratch/midway2/gaow/GIT/github/fine-mapping
</code></pre></div></div>

<h2 id="the-workflow">The workflow</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/finemapping_results_wrangler.ipynb -h
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>usage: sos run workflow/finemapping_results_wrangler.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see "sos run -h" for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  default
  consolidate
  update_variant_id
  update_coordinate

Global Workflow Options:
  --loci . (as path)
                        Loci file
  --ss-data-prefix . (as path)
                        summary statistics file prefix
  --pattern 'uniform.SuSiE_B.L_5.prior_0p005.res_var_false'
                        identifier for fine-mapping results to be extracted

Sections
  default:
  consolidate:          Consolidate fine-mapping results to a single file
    Workflow Options:
      --pip-thresh 0.05 (as float)
                        Keep PIP above these thresholds
      --round-off 6 (as int)
                        Round PIP to given digits
  update_variant_id:    Update variant ID based on chrom and pos, from per chrom
                        files (optional)
    Workflow Options:
      --id-map-prefix . (as path)
                        Path containing files for variant ID update rule Each
                        file is a separate chromsome
      --id-map-suffix bim
      --columns 2 4 (as list)
                        columns first element for variant ID, 2nd element for
                        genomic position is [2,4] for BIM files
      --chroms 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 (as list)
                        chromosome identifiers
  update_coordinate:    Update genomic coordinates based on variant ID, from per
                        chrom files (optional)
    Workflow Options:
      --coordinate-map-prefix . (as path)
                        Path containing files for coordinate update rule Each
                        file is a separate chromsome
      --coordinate-map-suffix bim
      --coordinate-version-id hgX
                        coordinate identifier
      --columns 2 4 5 6 (as list)
                        columns first element for variant ID, 2nd element for
                        genomic position 3rd for reference allele and 4th for
                        alternative allele is [2,4,5,6] for BIM files
      --chroms 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 (as list)
                        chromosome identifiers

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[global]
# Loci file
parameter: loci = path()
# summary statistics file prefix
parameter: ss_data_prefix = path()
# identifier for fine-mapping results to be extracted
parameter: pattern = "uniform.SuSiE_B.L_5.prior_0p005.res_var_false"

fail_if(not loci.is_file(), msg = 'Please specify valid path for --loci')
fail_if(ss_data_prefix.is_file(), msg = '--ss-data-prefix should be a path not a file (usually file without extension, if using input from my data wrangling pipeline)')

ss_data_prefix = ss_data_prefix.absolute()
chunks = [x.strip().split() for x in open(f'{loci:a}').readlines() if not x.strip().startswith('#')]
chunks = [x[3] if len(x) == 4 else "%s_%s_%s" % (x[0], x[1], x[2]) for x in chunks]
data = [f'{ss_data_prefix}/{x}/{x}.summary_stats.gz' for x in chunks]


if 'SuSiE' in pattern:
    source = 'susie'
elif 'CAVIAR' in pattern:
    source = 'caviar'
elif 'DAP' in pattern:
    source = 'dap'
elif 'FINEMAP' in pattern:
    source = 'finemap'
else:
    raise ValueError("Invalid --pattern specification")
</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[default]
sos_run('consolidate+update_variant_id+update_coordinate')
</code></pre></div></div>

<h2 id="fine-mapping-results-consolidation">Fine-mapping results consolidation</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Consolidate fine-mapping results to a single file
[consolidate]
depends: R_library('rlang&gt;=0.3.0'), R_library('dplyr'), R_library('data.table'), R_library('R.utils'), R_library('dscrutils') # can be installed via `devtools::install_github("stephenslab/dsc",subdir = "dscrutils", force = TRUE)`
# Keep PIP above these thresholds
parameter: pip_thresh = 0.05
# Round PIP to given digits
parameter: round_off = 6

ext = 'pkl' if source in ['dap'] else 'rds'

input: [x for x in paths([f'{ss_data_prefix}/{c}/{c}.{pattern}.{ext}' for c in chunks]) if x.is_file()]
output: f"{ss_data_prefix}.{pattern}.{loci:bn}.gz"
fail_if(len(_input) == 0, msg = f'Cannot find valid input files for given loci')
R: expand = "${ }", workdir = ss_data_prefix, stdout = f'{_output:n}.log'
    # Here we define get_*_output functions for different output format
    get_susie_output = function(unit, rds_file) {
        cs_id = cs_size = cs_purity = rep(NA, length(rds_file$var_names))
        num_cs = length(rds_file$sets$cs)
        for(id in 1:num_cs){
            idx = rds_file$sets$cs[[id]]
            cs_id[idx] = names(rds_file$sets$cs)[id]
            cs_size[idx] = length(rds_file$sets$cs[[id]])
            cs_purity[idx] = rds_file$sets$purity[id,1]
        }
        out = cbind.data.frame(rep(unit, length(rds_file$var_names)),
                                rds_file$var_names,
                                rds_file$pip, cs_id, cs_size, cs_purity)
        colnames(out) = c("locus_id", "variant_id", "pip", "cs", "cs_size", "cs_purity")
        out[which(out[,3] &gt;= ${pip_thresh} | !is.na(out[,4])), ]
    }
   
    get_caviar_output = function(unit, rds_file) {
  
      cs_annot = rep(0, length(rds_file$var_names))
      cs_annot[match(rds_file$set, rds_file$snp$snp)] = 1 
      out = cbind.data.frame(rep(unit, length(rds_file$var_names)),
                              rds_file$var_names,
                              rds_file$snp$snp_prob,
                              cs_annot)
      colnames(out) = c("gene_id", "var_id", "pip", "cs")
      out[which(out[,3] &gt;= ${pip_thresh}), ]
    }

    get_dap_output = function(unit, rds_file) {
      out = cbind.data.frame(rep(unit, length(rds_file$var_names)),
                              rds_file$var_names,
                              rds_file$snp$snp_prob,
                              rds_file$snp$cluster)
      colnames(out) = c("gene_id", "var_id", "pip", "cs")
      out[which(out[,3] &gt;= ${pip_thresh} | out[,4] &gt; 0),]
    }  
  
    get_finemap_output = get_caviar_output
  
    is_float = function(x) {is.numeric(x) &amp;&amp; !is.integer(x)}

    # Data extraction script
    suppressMessages(library(dplyr))
    suppressMessages(library(data.table))
    files = c(${paths([x.relative_to(ss_data_prefix) for x in _input]):r,})
    processed_dat = list()
    idx = 0
    for (f in files) {
      idx = idx + 1
      rds_file = dscrutils::read_dsc(f)
      unit = dirname(f)
      processed_dat[[idx]] = get_${source}_output(unit, rds_file)
      cat("We are at unit #", idx, ":", unit, "\n")
    }
    processed_dat = data.frame(rbindlist(processed_dat))
    extract_coord = data.frame(do.call(rbind, lapply(processed_dat[,2], function(x) return(strsplit(as.character(x), ":")[[1]]))))
    df = data.frame("chr" = extract_coord[,1], "pos" = extract_coord[,2], "ref" = extract_coord[,3], "alt" = extract_coord[,4],
                    "variant_id" = processed_dat[,2], "locus_id" = processed_dat[,1])
    df = cbind.data.frame(df, processed_dat[, 3:ncol(processed_dat)])
    df %&gt;% mutate_if(is_float, ~round(., ${round_off})) -&gt; df
    df_sorted = df [order(df[,1], df[,2]),]
    fwrite(df_sorted, file = "${_output:n}.tmp", sep = "\t", quote=FALSE, row.names=FALSE)
    R.utils::gzip("${_output:n}.tmp", destname=${_output:r}, overwrite=TRUE, remove=TRUE)
</code></pre></div></div>

<h2 id="update-variant-id-via-genomic-coordinantes">Update variant ID via genomic coordinantes</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Update variant ID based on chrom and pos, from per chrom files (optional)
[update_variant_id]
depends: R_library('rlang&gt;=0.3.0'), R_library('dplyr'), R_library('R.utils'), R_library('data.table')
# Path containing files for variant ID update rule
# Each file is a separate chromsome
parameter: id_map_prefix = path()
parameter: id_map_suffix = 'bim'
# columns first element for variant ID, 2nd element for genomic position
# is [2,4] for BIM files
parameter: columns = [2,4]
# chromosome identifiers
parameter: chroms = [x+1 for x in range(22)]
skip_if(len(glob.glob(f"{id_map_prefix:a}.*.{id_map_suffix}")) == 0, msg = 'Variant ID are not updated because no valid file is found using --id-map-prefix and --id-map-suffix')
output: f'{_input:n}.id_updated.gz'
R: expand = "${ }", stdout = f'{_output:n}.log'
    suppressMessages(library(dplyr))
    suppressMessages(library(data.table))
    out =  data.frame(fread("zcat ${_input}"))
    out %&gt;% mutate_if(is.factor, as.character) -&gt; out
    chroms = c(${paths(chroms):r,})
    for(numchr in chroms){
        which_chr = which(out$chr == numchr)
        out_sub = out[which_chr, ]
        dbfile = data.frame(fread(paste0("${id_map_prefix}.", numchr, ".${id_map_suffix}")))
        out_sub_new = out_sub
        idx1 = match(out_sub$pos, dbfile$V${columns[1]})
        idx2 = idx1[which(!is.na(idx1))]
        idx3 = 1:length(out_sub$pos)
        idx4 = idx3[which(!is.na(idx1))]
        out_sub_new[idx4, "variant_id"] = dbfile[idx2,"V${columns[0]}"]
        out[which_chr, ] = out_sub_new
        cat("Variant IDs updated for chromosome", numchr, "\n")
    }
    fwrite(out, file = "${_output:n}.tmp", sep = "\t", quote=FALSE, row.names=FALSE)
    R.utils::gzip("${_output:n}.tmp", destname=${_output:r}, overwrite=TRUE, remove=TRUE)
</code></pre></div></div>

<h2 id="update-genomic-coordinates-via-variant-id">Update genomic coordinates via variant ID</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Update genomic coordinates based on variant ID, from per chrom files (optional)
[update_coordinate]
depends: R_library('rlang&gt;=0.3.0'), R_library('dplyr'), R_library('data.table'), R_library('R.utils')
# Path containing files for coordinate update rule
# Each file is a separate chromsome
parameter: coordinate_map_prefix = path()
parameter: coordinate_map_suffix = 'bim'
# coordinate identifier
parameter: coordinate_version_id = 'hgX'
# columns first element for variant ID, 2nd element for genomic position
# 3rd for reference allele and 4th for alternative allele
# is [2,4,5,6] for BIM files
parameter: columns = [2,4,5,6]
# chromosome identifiers
parameter: chroms = [x+1 for x in range(22)]
skip_if(len(glob.glob(f"{coordinate_map_prefix:a}.*.{coordinate_map_suffix}")) == 0, msg = 'Genomic coordinates are not updated because no valid file is found using --coordinate-map-prefix and --coordinate-map-suffix')
output: f'{_input:n}.{coordinate_version_id}.gz'.replace('.id_updated.', '.')
R: expand = "${ }", stdout = f'{_output:n}.log'
    suppressMessages(library(dplyr))
    suppressMessages(library(data.table))
    out =  data.frame(fread("zcat ${_input}"))
    out %&gt;% mutate_if(is.factor, as.character) -&gt; out
    chroms = c(${paths(chroms):r,})
    out2 = c()
    for(numchr in chroms){
        dbfile = data.frame(fread(paste0("${coordinate_map_prefix}.", numchr, ".${coordinate_map_suffix}")))
        out2_sub = out[which(out$chr == numchr), ]
        idx1 = match(out2_sub$variant_id, dbfile$V${columns[0]})
        idx2 = idx1[which(!is.na(idx1))]
        idx3 = 1:length(out2_sub$variant_id)
        idx4 = idx3[which(!is.na(idx1))]
        out2_sub[idx4, c("pos", "ref", "alt")] = dbfile[idx2, c("V${columns[1]}", "V${columns[2]}", "V${columns[3]}")]
        out2 = rbind.data.frame(out2, out2_sub[idx4,])
        cat("Genomic coordinate updated for chromosome", numchr, "\n")
    }
    fwrite(out2, file = "${_output:n}.tmp", sep = "\t", quote=FALSE, row.names=FALSE)
    R.utils::gzip("${_output:n}.tmp", destname=${_output:r}, overwrite=TRUE, remove=TRUE)
</code></pre></div></div>

<h2 id="example-run">Example run</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/finemapping_results_wrangler.ipynb \
    --ss-data-prefix ~/tmp/01-Jan-2019 \
    --pattern uniform.SuSiE_B.L_5.prior_0p005.res_var_false \
    --id-map-prefix /project2/mstephens/SuSIE_gtex_CPP/rsID_map/Hg38/1000G.EUR.hg38 \
    --coordinate-map-prefix /project2/mstephens/SuSIE_gtex_CPP/rsID_map/Hg37/1000G.EUR.QC \
    --coordinate-version-id hg37
</code></pre></div></div>


              <nav class="c-page__nav">
  
    
    <a id="js-page__nav__prev" class="c-page__nav__prev" href="/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping">
      〈 <span class="u-margin-right-tiny"></span> DSC configuration syntax
    </a>
  

  
    
    <a id="js-page__nav__next" class="c-page__nav__next" href="">
       <span class="u-margin-right-tiny"></span> 〉
    </a>
  
</nav>

            </div>
          </div>
        </div>
      </main>
    </div>

  </body>
</html>
