<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width,minimum-scale=1">

  <title>DSC configuration syntax</title>
  <meta name="description" content="Fine-mapping workflow using summary statistics">

  <link rel="canonical" href="https://gaow.github.io/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping.html">
  <link rel="alternate" type="application/rss+xml" title="Gao's recipe" href="https://gaow.github.io/jupyterbook-sos-demo/feed.xml">

  <meta property="og:url"         content="https://gaow.github.io/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="DSC configuration syntax" />
<meta property="og:description" content="Fine-mapping workflow using summary statistics" />
<meta property="og:image"       content="" />


  <script type="application/ld+json">
  {
  "@context": "http://schema.org",
  "@type": "NewsArticle",
  "mainEntityOfPage":
    "https://gaow.github.io/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping.html",
  "headline":
    "DSC configuration syntax",
  "datePublished":
    "2019-02-01T10:03:57-06:00",
  "dateModified":
    "2019-02-01T10:03:57-06:00",
  "description":
    "Fine-mapping workflow using summary statistics",
  "author": {
    "@type": "Person",
    "name": "Gao Wang and others, Stephens Lab, The University of Chicago"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Data 100 at UC Berkeley",
    "logo": {
      "@type": "ImageObject",
      "url": "https://gaow.github.io/jupyterbook-sos-demo",
      "width": 60,
      "height": 60
    }
  },
  "image": {
    "@type": "ImageObject",
    "url": "https://gaow.github.io/jupyterbook-sos-demo",
    "height": 60,
    "width": 60
  }
}

  </script>
  <link rel="stylesheet" href="/jupyterbook-sos-demo/assets/css/styles.css">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css ">
  <link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
  <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
  <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
  <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
  <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
  <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
  <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
  <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">

  <!-- <link rel="manifest" href="/manifest.json"> -->
  <!-- <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#efae0a"> -->
  <meta name="msapplication-TileColor" content="#da532c">
  <meta name="msapplication-TileImage" content="/mstile-144x144.png">
  <meta name="theme-color" content="#233947">

  <!-- Favicon -->
  <link rel="shortcut icon" type="image/x-icon" href="/jupyterbook-sos-demo/images/favicon.ico">

  <!-- MathJax Config -->
  <!-- Allow inline math using $ and automatically break long math lines -->
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    CommonHTML: {
        linebreaks: {
            automatic: true,
        },
    },
});
</script>
<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_CHTML' async></script>

  <!-- DOM updating function -->
  <script>
const runWhenDOMLoaded = cb => {
  if (document.readyState != 'loading') {
    cb()
  } else if (document.addEventListener) {
    document.addEventListener('DOMContentLoaded', cb)
  } else {
    document.attachEvent('onreadystatechange', function() {
      if (document.readyState == 'complete') cb()
    })
  }
}

// Helper function to init things quickly
initFunction = function(myfunc) {
  runWhenDOMLoaded(myfunc);
  document.addEventListener('turbolinks:load', myfunc);
};
</script>

  <!-- Define some javascript variables that will be useful in other javascript -->
  <script>
    const site_basename = '/jupyterbook-sos-demo';
  </script>

  <!-- Add AnchorJS to let headers be linked -->
  <script src="/jupyterbook-sos-demo/assets/js/anchor.min.js"  type="text/javascript"></script>
  <script>

initFunction(function () {
    anchors.add("main h1, main h2, main h3, main h4")
});

</script>

  <!-- Include Turbolinks to make page loads fast -->
  <!-- https://github.com/turbolinks/turbolinks -->
  <script src="/jupyterbook-sos-demo/assets/js/turbolinks.js" async></script>
  <meta name="turbolinks-cache-control" content="no-cache">

  <!-- Load nbinteract for widgets -->
  <script src="https://unpkg.com/nbinteract-core" async></script>

  <!-- Load Thebelab for interactive widgets -->
  <!-- Include Thebelab for interactive code if it's enabled -->


  <!-- Google analytics -->
  <script src="/jupyterbook-sos-demo/assets/js/ga.js" async></script>

  <!-- Clipboard copy button -->
  <script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js" async></script>

  <!-- Load JS that depends on site variables -->
  <script>
/**
 * Set up copy/paste for code blocks
 */
const codeCellId = index => `codecell${index}`

const clipboardButton = id =>
  `<a class="btn copybtn o-tooltip--left" data-tooltip="Copy" data-clipboard-target="#${id}">
    <img src="/jupyterbook-sos-demo/assets/images/copy-button.svg" alt="Copy to clipboard">
  </a>`

// Clears selected text since ClipboardJS will select the text when copying
const clearSelection = () => {
  if (window.getSelection) {
    window.getSelection().removeAllRanges()
  } else if (document.selection) {
    document.selection.empty()
  }
}

// Changes tooltip text for two seconds, then changes it back
const temporarilyChangeTooltip = (el, newText) => {
  const oldText = el.getAttribute('data-tooltip')
  el.setAttribute('data-tooltip', newText)
  setTimeout(() => el.setAttribute('data-tooltip', oldText), 2000)
}

const addCopyButtonToCodeCells = () => {
  // If ClipboardJS hasn't loaded, wait a bit and try again. This
  // happens because we load ClipboardJS asynchronously.
  if (window.ClipboardJS === undefined) {
    setTimeout(addCopyButtonToCodeCells, 250)
    return
  }

  const codeCells = document.querySelectorAll('.input_area pre')
  codeCells.forEach((codeCell, index) => {
    const id = codeCellId(index)
    codeCell.setAttribute('id', id)
    if (document.querySelector(`pre#${id} + a`) == null) {
      codeCell.insertAdjacentHTML('afterend', clipboardButton(id));
    }
  })

  const clipboard = new ClipboardJS('.copybtn')
  clipboard.on('success', event => {
    clearSelection()
    temporarilyChangeTooltip(event.trigger, 'Copied!')
  })

  clipboard.on('error', event => {
    temporarilyChangeTooltip(event.trigger, 'Failed to copy')
  })

  // Get rid of clipboard before the next page visit to avoid memory leak
  document.addEventListener('turbolinks:before-visit', () =>
    clipboard.destroy()
  )
}

initFunction(addCopyButtonToCodeCells);
</script>

  <!-- Hide cell code -->
  
<script>
/**
Add buttons to hide code cells
*/


var setCodeCellVisibility = function(inputField, kind) {
    // Update the image and class for hidden
    var id = inputField.getAttribute('data-id');
    var codeCell = document.querySelector(`#${id}`);

    if (kind === "visible") {
        codeCell.classList.remove('hidden');
        inputField.checked = true;
    } else {
        codeCell.classList.add('hidden');
        inputField.checked = false;
    }
}

var toggleCodeCellVisibility = function (event) {
    // The label is clicked, and now we decide what to do based on the input field's clicked status
    if (event.target.tagName === "LABEL") {
        var inputField = event.target.previousElementSibling;
    } else {
        // It is the span inside the target
        var inputField = event.target.parentElement.previousElementSibling;
    }

    if (inputField.checked === true) {
        setCodeCellVisibility(inputField, "visible");
    } else {
        setCodeCellVisibility(inputField, "hidden");
    }
}


// Button constructor
const hideCodeButton = id => `<input class="hidebtn" type="checkbox" id="hidebtn${id}" data-id="${id}"><label title="Toggle cell" for="hidebtn${id}" class="plusminus"><span class="pm_h"></span><span class="pm_v"></span></label>`

var addHideButton = function () {
  // If a hide button is already added, don't add another
  if (document.querySelector('div.hidecode input') !== null) {
      return;
  }

  // Find the input cells and add a hide button
  document.querySelectorAll('div.input_area div.highlight').forEach(function (item, index) {
    if (!item.parentElement.classList.contains("hidecode")) {
        // Skip the cell if it doesn't have a hidecode class
        return;
    }

    const id = codeCellId(index)
    item.setAttribute('id', id);
    item.insertAdjacentHTML('afterend', hideCodeButton(id))

    // Set up the visibility toggle
    hideLink = document.querySelector(`#${id} + input + label`);
    hideLink.addEventListener('click', toggleCodeCellVisibility)
  });
}


// Initialize the hide buttos
var initHiddenCells = function () {
    // Add hide buttons to the cells
    addHideButton();

    // Toggle the code cells that should be hidden
    document.querySelectorAll('div.hidecode input').forEach(function (item) {
        setCodeCellVisibility(item, 'hidden');
        item.checked = true;
    })
}

initFunction(initHiddenCells);

</script>


  <!-- Load custom website scripts -->
  <script src="/jupyterbook-sos-demo/assets/js/scripts.js" async></script>

  <!-- Load custom user CSS and JS  -->
  <script src="/jupyterbook-sos-demo/assets/custom/custom.js" async></script>
  <link rel="stylesheet" href="/jupyterbook-sos-demo/assets/custom/custom.css">

  <!-- Update interact links w/ REST param, is defined in includes so we can use templates -->
  

  <!-- Lunr search code - will only be executed on the /search page -->
  <script src="/jupyterbook-sos-demo/assets/js/lunr/lunr.min.js" type="text/javascript"></script>
  <script>var initQuery = function() {
  // See if we have a search box
  var searchInput = document.querySelector('input#lunr_search');
  if (searchInput === null) {
    return;
  }

  // Function to parse our lunr cache
  var idx = lunr(function () {
    this.field('title')
    this.field('excerpt')
    this.field('categories')
    this.field('tags')
    this.ref('id')

    this.pipeline.remove(lunr.trimmer)

    for (var item in store) {
      this.add({
        title: store[item].title,
        excerpt: store[item].excerpt,
        categories: store[item].categories,
        tags: store[item].tags,
        id: item
      })
    }
  });

  // Run search upon keyup
  searchInput.addEventListener('keyup', function () {
    var resultdiv = document.querySelector('#results');
    var query = document.querySelector("input#lunr_search").value.toLowerCase();
    var result =
      idx.query(function (q) {
        query.split(lunr.tokenizer.separator).forEach(function (term) {
          q.term(term, { boost: 100 })
          if(query.lastIndexOf(" ") != query.length-1){
            q.term(term, {  usePipeline: false, wildcard: lunr.Query.wildcard.TRAILING, boost: 10 })
          }
          if (term != ""){
            q.term(term, {  usePipeline: false, editDistance: 1, boost: 1 })
          }
        })
      });

      // Empty the results div
      while (resultdiv.firstChild) {
        resultdiv.removeChild(resultdiv.firstChild);
      }

    resultdiv.insertAdjacentHTML('afterbegin', '<p class="results__found">'+result.length+' Result(s) found</p>');
    for (var item in result) {
      var ref = result[item].ref;
      if(store[ref].teaser){
        var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<div class="archive__item-teaser">'+
                '<img src="'+store[ref].teaser+'" alt="">'+
              '</div>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      else{
    	  var searchitem =
          '<div class="list__item">'+
            '<article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">'+
              '<h2 class="archive__item-title" itemprop="headline">'+
                '<a href="'+store[ref].url+'" rel="permalink">'+store[ref].title+'</a>'+
              '</h2>'+
              '<p class="archive__item-excerpt" itemprop="description">'+store[ref].excerpt.split(" ").splice(0,20).join(" ")+'...</p>'+
            '</article>'+
          '</div>';
      }
      resultdiv.insertAdjacentHTML('beforeend', searchitem);
    }
  });
};

initFunction(initQuery);
</script>
</head>

  <body>
    <!-- .js-show-sidebar shows sidebar by default -->
    <div id="js-textbook" class="c-textbook js-show-sidebar">
      



<nav id="js-sidebar" class="c-textbook__sidebar">
  <a href="https://github.com/gaow/jupyterbook-sos-demo"><img src="/jupyterbook-sos-demo/images/logo.svg" class="textbook_logo" id="sidebar-logo" data-turbolinks-permanent/></a>
  <h2 class="c-sidebar__title">Gao's recipe</h2>
  <ul class="c-sidebar__chapters">
    
      
      
        <li class="c-sidebar__chapter"><a class="c-sidebar__entry" href="/jupyterbook-sos-demo/search.html"><img src="/jupyterbook-sos-demo/images/search.svg" alt="Search This Wiki"></a></li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/jupyterbook-sos-demo/README.html"
        >
          
          Home
        </a>

        
      </li>

      
    
      
      
        <li class="c-sidebar__divider"></li>
        
      
      
        <li><h2 class="c-sidebar__title">Workflows</li>
        
      
      

      
      

      
      
      <li class="c-sidebar__chapter">
        <a class="c-sidebar__entry "
          href="/jupyterbook-sos-demo/finemapping/README.html"
        >
          
          Genetic fine-mapping
        </a>

        

          
          
          
          

          

          <ul class="c-sidebar__sections ">
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry "
                  href="/jupyterbook-sos-demo/finemapping/summary_statistics_wrangler.html"
                >
                  
                  Matching summary data and reference panel
                </a>

                
                

              </li>
              
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry c-sidebar__entry--active"
                  href="/jupyterbook-sos-demo/finemapping/summary_statistics_finemapping.html"
                >
                  
                  Fine-mapping with summary data
                </a>

                
                

              </li>
              
            
              
              

              <li class="c-sidebar__section">
                <a class="c-sidebar__entry "
                  href="/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler.html"
                >
                  
                  Consolidating genome-wide results
                </a>

                
                

              </li>
              
            
          </ul>
        
      </li>

      
    
  </ul>
  <p class="sidebar_footer">Copyright (c) 2019 <br> <a href="http://home.uchicago.edu/gaow">Gao Wang</a> and colleagues<br></p>
</nav>

      
      <!-- Shamelessly copied from minimal mistakes -->


<!-- TOC will only show up if it has at least one item -->


  <aside class="sidebar__right">
    <nav class="onthispage">
      <header><h4 class="nav__title"><i class="fa fa-list"></i>   Contents</h4></header>
      <ul class="toc__menu">
  <li><a href="#software-preparation">Software preparation</a></li>
  <li><a href="#input-data">Input data</a></li>
  <li><a href="#output-format">Output format</a></li>
  <li><a href="#the-workflow">The workflow</a>
    <ul>
      <li><a href="#susie-hatb-version">SuSiE $\hat{b}$ version</a></li>
      <li><a href="#susie-z-version">SuSiE $z$ version</a></li>
      <li><a href="#dap-g">DAP-G</a></li>
      <li><a href="#caviar">CAVIAR</a></li>
      <li><a href="#finemap">FINEMAP</a></li>
      <li><a href="#make-diagnostic-plot">Make diagnostic plot</a></li>
    </ul>
  </li>
  <li><a href="#example-usage">Example usage</a>
    <ul>
      <li><a href="#susie-hatb">SuSiE $\hat{b}$</a></li>
      <li><a href="#susie-z">SuSiE $z$</a></li>
      <li><a href="#dap">DAP</a></li>
      <li><a href="#caviar-1">CAVIAR</a></li>
      <li><a href="#finemap-1">FINEMAP</a></li>
    </ul>
  </li>
</ul>
    </nav>
  </aside>


      
      <main class="c-textbook__page" tabindex="-1">
          <div class="o-wrapper">
            <div class="c-sidebar-toggle">
  <!-- We show the sidebar by default so we use .is-active -->
  <button
    id="js-sidebar-toggle"
    class="hamburger hamburger--arrowalt is-active"
  >
    <span class="hamburger-box">
      <span class="hamburger-inner"></span>
    </span>
    <span class="c-sidebar-toggle__label">Toggle Sidebar</span>
  </button>
</div>

            
<div class="buttons">




</div>


            <div class="c-textbook__content">
              <h1 id="fine-mapping-workflow-using-summary-statistics">Fine-mapping workflow using summary statistics</h1>

<p>This is a workflow that does summary statistics fine-mapping using SuSiE, DAP-G and CAVIAR.</p>

<p>This pipeline was devised by Gao Wang and implemented by Gao Wang with input from Yuxin Zou at The University of Chicago. It can be downloaded <a href="https://github.com/gaow/fine-mapping/tree/master/workflow">from here</a>. 
SuSiE is implemented in <a href="https://github.com/stephenslab/susieR"><code class="highlighter-rouge">susieR</code></a> (please use version 0.6.2+).</p>

<p><code class="highlighter-rouge">susieR</code> is flexible in input format (R data matrices). Since we need to analyze the data with command executables such as DAP-G and CAVIAR we make input format compatible with these tools.</p>

<h2 id="software-preparation">Software preparation</h2>

<p>Update SoS and sos-notebook</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip install -U sos sos-notebook
</code></pre></div></div>

<p>To run SuSiE, Install <code class="highlighter-rouge">susieR</code> and <code class="highlighter-rouge">dscrutils</code></p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>devtools::install_github("stephenslab/susieR")
devtools::install_github("stephenslab/dsc", subdir = "dscrutils", force = TRUE)
</code></pre></div></div>

<p>To run GAP-G and CAVIAR, make sure GSL exists</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sudo apt-get install libgsl-dev
</code></pre></div></div>

<p>and make sure DAP-G and CAVIAR exists under <code class="highlighter-rouge">$PATH</code>.</p>

<p>To install a specified version of CAVIAR, eg for commit ID <code class="highlighter-rouge">743038a</code>:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>curl -L https://github.com/fhormoz/caviar/tarball/743038a32ae66ea06ee599670cb7939fb80a923f -o caviar.tar.gz \
    &amp;&amp; tar -zxvf caviar.tar.gz &amp;&amp; cd fhormoz-caviar-*/CAVIAR-C++ &amp;&amp; make
</code></pre></div></div>

<p>You should then see files <code class="highlighter-rouge">CAVIAR eCAVIAR mupCAVIAR setCAVIAR</code> under the current directory <code class="highlighter-rouge">fhormoz-caviar-*/CAVIAR-C++</code>. You can either move them to your system’s <code class="highlighter-rouge">$PATH</code>, or just leave them there and use <code class="highlighter-rouge">-b</code> option in <code class="highlighter-rouge">sos run</code> call, see examples in workflow demo commands below.</p>

<p>Same for DAP-G,</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>curl -L https://github.com/xqwen/dap/tarball/ef11b263ae5e11b9e2e295757927877c03274095 -o dap.tar.gz \
    &amp;&amp; tar -zxvf dap.tar.gz &amp;&amp; cd xqwen-dap-*/dap_src &amp;&amp; make
</code></pre></div></div>

<p>You should see file <code class="highlighter-rouge">dap-g</code> under <code class="highlighter-rouge">xqwen-dap-*/dap_src</code>.</p>

<h2 id="input-data">Input data</h2>

<p>Input data relies on the output of the pipeline <code class="highlighter-rouge">summary_statistics_wrangler.ipynb</code> (an example input data-set for this pipeline can be found in that notebook). The input format is the same for SuSiE, DAP and CAVIAR.</p>

<h2 id="output-format">Output format</h2>

<p>Output is still work in progress.</p>

<p>Currently they are just R or Python data objects <code class="highlighter-rouge">*.rds</code> or <code class="highlighter-rouge">*.pkl</code> files and can be loaded via <code class="highlighter-rouge">dscrutils::read_dsc</code>. It contains the most comprehensive information and can be used to generate more succinct output files downstream. Additionally there will be a PDF file to visualize PIP and original $z$ scores.</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%cd ~/GIT/github/fine-mapping
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/scratch/midway2/gaow/GIT/github/fine-mapping
</code></pre></div></div>

<h2 id="the-workflow">The workflow</h2>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/summary_statistics_finemapping.ipynb -h
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>usage: sos run workflow/summary_statistics_finemapping.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see "sos run -h" for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  susie_bhat
  susie_z
  dap
  caviar
  finemap

Global Workflow Options:
  --loci . (as path)
                        Loci file
  --ss-data-prefix . (as path)
                        summary statistics file prefix
  --annotation uniform
                        annotation "tag"
  --n 0 (as int)
                        sample size
  --var-y 0 (as int)
                        variance of phenotype. Set to zero for "unspecified"
  --job-size 80 (as int)
                        For cluster jobs, number of loci to analyze per job

Sections
  susie_bhat_1:         Run SuSiE bhat version
    Workflow Options:
      --susie-pve 0.1 (as float)
                        Set pve to 0.0 to estimate from data
      --[no-]pve-is-total (default to False)
                        use --pve-is-total to indicate the pve set above is for
                        the entire loci and should be divided over number of
                        variables
      --[no-]susie-est-residual (default to False)
                        Whether or not to estimate residual variance
      --L 5 (as int)
                        Set maximum number of effect variables
      --[no-]track-fit (default to False)
                        Whether or not to keep fit from each iteration, for
                        debug purpose.
  susie_z_1:            Run SuSiE Z score version
    Workflow Options:
      --L 5 (as int)
                        Set maximum number of effect variables
      --[no-]track-fit (default to False)
                        Whether or not to keep fit from each iteration, for
                        debug purpose.
  dap_1:                Run DAP-G
    Workflow Options:
      --args ''
                        Extra arguments to pass to DAP
  caviar_1:             Run CAVIAR
    Workflow Options:
      --args ''
                        Extra arguments to pass to CAVIAR
  finemap_1:            Run FINEMAP
    Workflow Options:
      --k NULL
                        K parameter
      --args ''
                        Extra arguments to pass to CAVIAR
  *_2:                  Summarize Fine-mapping results

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[global]
# Loci file
parameter: loci = path()
# summary statistics file prefix
parameter: ss_data_prefix = path()
# annotation "tag"
parameter: annotation = "uniform"
# sample size
parameter: n = 0
# variance of phenotype. Set to zero for "unspecified"
parameter: var_y = 0
# For cluster jobs, number of loci to analyze per job
parameter: job_size = 80

fail_if(not loci.is_file(), msg = 'Please specify valid path for --loci')
fail_if(ss_data_prefix.is_file(), msg = '--ss-data-prefix should be a path not a file (usually file without extension, if using input from my data wrangling pipeline)')

ss_data_prefix = ss_data_prefix.absolute()
chunks = [x.strip().split() for x in open(f'{loci:a}').readlines() if not x.strip().startswith('#')]
chunks = [x[3] if len(x) == 4 else "%s_%s_%s" % (x[0], x[1], x[2]) for x in chunks]
data = [f'{ss_data_prefix}/{x}/{x}.summary_stats.gz' for x in chunks]

def fmtP(x):
    return str(x).replace(".", "p").replace(' ', '_').replace('"', "").replace("'", "").replace("-", '_')

def is_data_file_empty(f, min_row = 0):
    if f.stat().st_size == 0:
        return True
    import pandas
    try:
        x = pandas.read_csv(str(f), compression='gzip', nrows=min_row + 1, delim_whitespace = True)
    except pandas.errors.EmptyDataError:
        return True
    if x.shape[1] == 1:
        return True
    if x.shape[0] &lt;= min_row:
        return True
    return False
</code></pre></div></div>

<h3 id="susie-hatb-version">SuSiE $\hat{b}$ version</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Run SuSiE bhat version
[susie_bhat_1]
fail_if(n&lt;=2, msg = 'Please specify valid sample size --n')
depends: R_library('data.table'), R_library('susieR')
# Set pve to 0.0 to estimate from data
parameter: susie_pve = 0.1
# use --pve-is-total to indicate the pve set above is for the entire loci
# and should be divided over number of variables
parameter: pve_is_total = False
# Whether or not to estimate residual variance
parameter: susie_est_residual = False
# Set maximum number of effect variables
parameter: L = 5
# Whether or not to keep fit from each iteration,
# for debug purpose.
parameter: track_fit = False
#
suffix = f'SuSiE_B.L_{L}.prior_{fmtP(susie_pve)}.res_var_{str(susie_est_residual).lower()}'
var_y_str = f'var_y = {var_y},' if var_y &gt; 0 else ''
#
input: data, group_by = 1
output: f'{_input:nn}.{annotation}.{suffix}.rds'
skip_if(is_data_file_empty(_input))

task: trunk_workers = 1, trunk_size = job_size, walltime = '10m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'

prior_str = f'read.table(gzfile("{_input:nn}.{annotation}.gz"))[,2]' if os.path.isfile(f"{_input:nn}.{annotation}.gz") else 'NULL'
fail_if(annotation != 'uniform' and prior_str == 'NULL', msg = f"Cannot find annotation file for ``{annotation}``. Please use --annotation uniform if you do not want to apply any annotations")
R: expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'
    data = read.table(gzfile(${_input:ar}), head = F, stringsAsFactors = FALSE)
    prior = ${prior_str}
    r = as.matrix(data.table::fread("zcat ${_input:nn}.LD.gz"))
    if (ncol(data) == 3) {
      bhat = data[,2]
      shat = data[,3]
      p_values = 2 * pt(-abs(bhat / shat), df = ${n} - 2)
      z_scores = abs(qnorm(p_values/2)) * sign(bhat)      
    } else {
      bhat = data[,2]
      shat = rep(1, length(bhat))
      z_scores = data[,2]
    }
    library(susieR)
    L = min(${L}, length(z_scores))
    s = susie_bhat(bhat, shat, r, ${n}, ${var_y_str} r_tol = 5e-3, L = L, min_abs_corr = 0.1,
                  ${'estimate_prior_variance=TRUE' if susie_pve == 0 else 'scaled_prior_variance=' + str(susie_pve) + ('/length(bhat)' if pve_is_total else '')},
                  ${'estimate_residual_variance=TRUE' if susie_est_residual else 'estimate_residual_variance=FALSE'},
                  prior_weights=prior, track_fit=${"T" if track_fit else "F"})
    s$prior = prior
    s$z = z_scores
    s$var_names = data[,1]
    if (${susie_pve} == 0 &amp;&amp; ${int(susie_est_residual)}) print(paste(c("Estimated PVE:", round(s$V/s$sigma2, 4)), collapse = ' '))
    if (${"T" if track_fit else "F"}) susie_plot_iteration(s, L, ${_input:nr})
    saveRDS(s, ${_output:r})
</code></pre></div></div>

<h3 id="susie-z-version">SuSiE $z$ version</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Run SuSiE Z score version
[susie_z_1]
depends: R_library('data.table'), R_library('susieR')
# Set maximum number of effect variables
parameter: L = 5
# Whether or not to keep fit from each iteration,
# for debug purpose.
parameter: track_fit = False
#
suffix = f'SuSiE_Z.L_{L}'
#
input: data, group_by = 1
output: f'{_input:nn}.{annotation}.{suffix}.rds'
skip_if(is_data_file_empty(_input))

task: trunk_workers = 1, trunk_size = job_size, walltime = '10m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'

prior_str = f'read.table(gzfile("{_input:nn}.{annotation}.gz"))[,2]' if os.path.isfile(f"{_input:nn}.{annotation}.gz") else 'NULL'
fail_if(annotation != 'uniform' and prior_str == 'NULL', msg = f"Cannot find annotation file for ``{annotation}``. Please use --annotation uniform if you do not want to apply any annotations")
R: expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'
    data = read.table(gzfile(${_input:ar}), head = F, stringsAsFactors = FALSE)
    prior = ${prior_str}
    r = as.matrix(data.table::fread("zcat ${_input:nn}.LD.gz"))
    if (ncol(data) == 3) {
      bhat = data[,2]
      shat = data[,3]
      if (${n} &gt; 0) {
          p_values = 2 * pt(-abs(bhat / shat), df = ${n} - 2)
          z_scores = abs(qnorm(p_values/2)) * sign(bhat)
      } else {
          z_scores = data[,2] / data[,3]
      }
    } else {
      z_scores = data[,2]
    }
    library(susieR)
    L = min(${L}, length(z_scores))
    s = susie_z(z_scores, r, r_tol = 5e-3, L = L, min_abs_corr = 0.1,
                  prior_weights=prior, track_fit=${"T" if track_fit else "F"})
    s$prior = prior
    s$z = z_scores
    s$var_names = data[,1]
    if (${"T" if track_fit else "F"}) susie_plot_iteration(s, L, ${_input:nr})
    saveRDS(s, ${_output:r})
</code></pre></div></div>

<h3 id="dap-g">DAP-G</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Run DAP-G
[dap_1]
fail_if(n&lt;=2, msg = 'Please specif valid sample size --n')
# Extra arguments to pass to DAP
parameter: args = ''
suffix = f'DAP{"."+fmtP(args) if len(args) else ""}'
#
depends: executable('dap-g'), Py_Module('pandas'), Py_Module('numpy')
input: data, group_by = 1
output: f'{_input:nn}.{annotation}.{suffix}.pkl'
skip_if(is_data_file_empty(_input))

task: trunk_workers = 1, trunk_size = job_size, walltime = '10m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'
prior_file = f"{_input:ann}.{annotation}.gz" if os.path.isfile(f"{_input:ann}.{annotation}.gz") else None
fail_if(annotation != 'uniform' and prior_file is None, msg = f"Cannot find annotation file for ``{annotation}``. Please use --annotation uniform if you do not want to apply any annotations")

python: expand = '${ }', stderr = f'{_output:n}.stderr', workdir = f'{_output:ad}'
    import sys
    import pandas as pd
    import numpy as np
    import os, subprocess
    import pickle
    def run_dap_ss(prefix, data_file, ld_file, n, varY, prior_file = None, args = ''):
        cmd = ['dap-g', '-d_est', f'&lt;(zcat {data_file})', '-d_ld', f'&lt;(zcat {ld_file})', '-d_n', n, '-d_syy', varY, '-o', f'{prefix}.result', '-l', f'{prefix}.stderr', '--all', args]
        if prior_file is not None:
            cmd.extend(['-p', f'&lt;(zcat {prior_file})'])
        cmd = ' '.join(map(str,cmd))
        sys.stderr.write(cmd + '\n')
        subprocess.Popen(cmd, shell=True, executable='/bin/bash')

    def extract_dap_output(prefix):
        out = [x.strip().split() for x in open(f'{prefix}.result').readlines()]
        pips = []
        clusters = []
        still_pip = True
        for line in out:
            if len(line) == 0:
                continue
            if len(line) &gt; 2 and line[2] == 'cluster_pip':
                still_pip = False
                continue
            if still_pip and (not line[0].startswith('((')):
                continue
            if still_pip:
                pips.append([line[1], float(line[2]), float(line[3]), int(line[4])])
            else:
                clusters.append([len(clusters) + 1, float(line[2]), float(line[3])])
        pips = pd.DataFrame(pips, columns = ['snp', 'snp_prob', 'snp_log10bf', 'cluster'])
        clusters = pd.DataFrame(clusters, columns = ['cluster', 'cluster_prob', 'cluster_avg_r2'])
        clusters = pd.merge(clusters, pips.groupby(['cluster'])['snp'].apply(','.join).reset_index(), on = 'cluster')
        return {'snp': pips, 'set': clusters}


    prior_file = None if ${prior_file is None} else "${prior_file}"
    run_dap_ss(${_output:bnr}, ${_input:ar}, "${_input:nn}.LD.gz", ${n}, ${1 if var_y == 0 else var_y}, prior_file, ${args})
    res = extract_dap_output(${_output:anr})
    res['var_names'] = np.loadtxt(${_input:ar}, dtype = str, usecols=0)
    ss = np.loadtxt(${_input:ar}, dtype = str)
    if ss.shape[1] == 3:
        res['z'] = ss[:,1].astype(float)/ss[:,2].astype(float)
    else:
        res['z'] = ss[:,1].astype(float)
    if prior_file is not None:
        res['prior'] = np.loadtxt(prior_file, dtype = float, usecols=1)
    pickle.dump(res, open(${_output:ar}, 'wb'))
</code></pre></div></div>

<h3 id="caviar">CAVIAR</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Run CAVIAR
[caviar_1]
depends: executable('CAVIAR'), R_library('dplyr&gt;=0.7.7'), R_library('readr'), R_library('magrittr'), R_library('dscrutils') # can be installed via `devtools::install_github("stephenslab/dsc",subdir = "dscrutils", force = TRUE)`
# Extra arguments to pass to CAVIAR
parameter: args = ''
suffix = f'CAVIAR{"."+fmtP(args) if len(args) else ""}'
#
input: data, group_by = 1
output: f'{_input:nn}.{suffix}.rds'
skip_if(is_data_file_empty(_input))

task: trunk_workers = 1, trunk_size = job_size, walltime = '15m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'

R: expand = '${ }', stderr = f'{_output:n}.log', stdout = f'{_output:n}.log'
    library(dplyr)
    library(readr)
    library(magrittr)

    #' CAVIAR I/O
    prefix = ${_input:nnr}
    cfg = list( prefix=paste0(prefix,".CAVIAR"),
                z=paste0(prefix,".CAVIAR.z"),
                ld=paste0(prefix,".LD.txt"),
                ld_gz=paste0(prefix,".LD.gz"),
                set=paste0(prefix,".CAVIAR_set"),
                post=paste0(prefix,".CAVIAR_post"),
                log=paste0(prefix,".CAVIAR.log"))
    data = read.table(gzfile(${_input:ar}), head = F, stringsAsFactors = FALSE)
    if (ncol(data) == 3) {
        if (${n} &gt; 0) {
          p_values = 2 * pt(-abs(data[,2] / data[,3]), df = ${n} - 2)
          z_scores = abs(qnorm(p_values/2)) * sign(data[,2])
        } else {
          z_scores = data[,2] / data[,3]
        }
        data = cbind(data[,1], z_scores)
        write.table(data,cfg$z,quote=F,col.names=F,row.names=F)
    } else {
      z_scores = data[,2]
      system("ln -s ${_input} ${_input:nnr}.CAVIAR.z")
    }

    #' Run CAVIAR
    #' https://github.com/fhormoz/caviar

    run_caviar &lt;- function(cfg, args = "")
    {
      dscrutils::run_cmd(paste("zcat", cfg$ld_gz, "&gt;", cfg$ld), ferr = FALSE)
      cmd = paste("CAVIAR", "-z", cfg$z, "-l", cfg$ld, "-o", cfg$prefix, args)
      dscrutils::run_cmd(cmd)
      unlink(cfg$ld, force=TRUE)
      unlink(cfg$z, force=TRUE)
      if(!all(file.exists(cfg$post, cfg$set, cfg$log))) {
          stop("Cannot find one of the post, set, and log files")
      }

      log &lt;- readLines(cfg$log)

      # read output tables
      snp &lt;- read.delim(cfg$post)  
      stopifnot(ncol(snp) == 3)
      names(snp) &lt;- c("snp", "snp_prob_set", "snp_prob")
      snp$snp &lt;- as.character(snp$snp)
      snp &lt;- rank_snp(snp)

      # `set` of snps
      set &lt;- readLines(cfg$set)
      set_ordered &lt;- left_join(data_frame(snp = set), snp, by = "snp") %&gt;% 
        arrange(rank) %$% snp
      return(list(snp=snp, set=set_ordered))
    }

    rank_snp &lt;- function(snp) {
      snp &lt;- arrange(snp, -snp_prob) %&gt;%
        mutate(
            rank = seq(1, n()),
            snp_prob_cumsum = cumsum(snp_prob) / sum(snp_prob)) %&gt;%
        select(rank, snp, snp_prob, snp_prob_cumsum, snp_prob_set)
      return(snp)    
    }

    res = run_caviar(cfg, "${args}")
    res$z = z_scores
    res$var_names = data[,1]
    saveRDS(res, ${_output:ar})
</code></pre></div></div>

<h3 id="finemap">FINEMAP</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Run FINEMAP
[finemap_1]
depends: executable('finemap'), R_library('dplyr&gt;=0.7.7'), R_library('readr'), R_library('magrittr'), R_library('dscrutils') # can be installed via `devtools::install_github("stephenslab/dsc",subdir = "dscrutils", force = TRUE)`
# K parameter
parameter: k = 'NULL'
# Extra arguments to pass to CAVIAR
parameter: args = ''
suffix = f'FINEMAP{"."+fmtP(args) if len(args) else ""}'
fail_if(n &lt; 2, msg = 'Sample size is invalid. Please set it properly via --n.')
#
input: data, group_by = 1
output: f'{_input:nn}.{suffix}.rds'
skip_if(is_data_file_empty(_input, min_row = 1))

task: trunk_workers = 1, trunk_size = job_size, walltime = '15m', mem = '8G', cores = 1, tags = f'{step_name}_{_output:bn}'

R: expand = '${ }', stderr = f'{_output:n}.log', stdout = f'{_output:n}.log'
    library(dplyr)
    library(readr)
    library(magrittr)

    #' FINEMAP I/O
    prefix = ${_input:nnr}
    cfg = list( prefix=paste0(prefix,".FINEMAP"),
                z=paste0(prefix,".FINEMAP.z"),
                ld=paste0(prefix,".LD.ld"),
                ld_gz=paste0(prefix,".LD.gz"),
                config=paste0(prefix,".FINEMAP.config"),
                k=paste0(prefix,".FINEMAP.k"),
                snp=paste0(prefix,".FINEMAP.snp"),
                meta=paste0(prefix,".FINEMAP.master"),
                log=paste0(prefix,".FINEMAP.log"))
    data = read.table(gzfile(${_input:ar}), head = F, stringsAsFactors = FALSE)
    if (ncol(data) == 3) {
        if (${n} &gt; 0) {
          p_values = 2 * pt(-abs(data[,2] / data[,3]), df = ${n} - 2)
          z_scores = abs(qnorm(p_values/2)) * sign(data[,2])
        } else {
          z_scores = data[,2] / data[,3]
        }
        data = cbind(data[,1], z_scores)
        write.table(data,cfg$z,quote=F,col.names=F,row.names=F)
    } else {
      z_scores = data[,2]
      system("ln -s ${_input} ${_input:nnr}.FINEMAP.z")
    }
    if (!is.null(${k})) {
      write.table(t(${k}),cfg$k,quote=F,col.names=F,row.names=F)
      write("z;ld;snp;config;k;log;n-ind",file=cfg$meta)
      write(paste(cfg$z, cfg$ld, cfg$snp, cfg$config, cfg$k, cfg$log, ${n}, sep=";"),
        file=cfg$meta,append=TRUE)
    } else {
      write("z;ld;snp;config;log;n-ind",file=cfg$meta)
      write(paste(cfg$z, cfg$ld, cfg$snp, cfg$config, cfg$log, ${n}, sep=";"),
            file=cfg$meta,append=TRUE)
    }

    #' Run FINEMAP 1.1

    run_finemap &lt;- function(cfg, args = "")
    {
      dscrutils::run_cmd(paste("zcat", cfg$ld_gz, "&gt;", cfg$ld), ferr = FALSE)
      cmd = paste("finemap --sss --log", "--in-files", cfg$meta, args)
      dscrutils::run_cmd(cmd)
      unlink(cfg$ld, force=TRUE)
      unlink(cfg$z, force=TRUE)
      if(!all(file.exists(cfg$post, cfg$set, cfg$log))) {
          stop("Cannot find one of the post, set, and log files")
      }

      log &lt;- readLines(cfg$log)

      # read output tables
      snp = read.table(cfg$snp,header=TRUE,sep=" ")
      snp$snp = as.character(snp$snp)

      snp = rank_snp(snp)
      config = read.table(cfg$config,header=TRUE,sep=" ")

      # extract number of causal
      ncausal = finemap_extract_ncausal(cfg$log)
      return(list(snp=snp, set=config, ncausal=ncausal))
    }

    rank_snp &lt;- function(snp) {
      snp &lt;- arrange(snp, -snp_prob) %&gt;%
        mutate(
            rank = seq(1, n()),
            snp_prob_cumsum = cumsum(snp_prob) / sum(snp_prob)) %&gt;%
        select(rank, snp, snp_prob, snp_prob_cumsum, snp_log10bf)
      return(snp)    
    }
  
    finemap_extract_ncausal &lt;- function(logfile)
    {
      lines &lt;- grep("-&gt;", readLines(logfile), value = TRUE)
      lines &lt;- gsub("\\(|\\)|&gt;", "", lines)
      splits &lt;- strsplit(lines, "\\s+")
      tab &lt;- data.frame(
        ncausal_num = sapply(splits, function(x) as.integer(x[2])),
        ncausal_prob = sapply(splits, function(x) as.double(x[4])))
      tab &lt;- mutate(tab, type = ifelse(duplicated(ncausal_num), "post", "prior"))
      return(tab)
    }

    res = run_finemap(cfg, "${args}")
    res$z = z_scores
    res$var_names = data[,1]
    saveRDS(res, ${_output:ar})
</code></pre></div></div>

<h3 id="make-diagnostic-plot">Make diagnostic plot</h3>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Summarize Fine-mapping results
[*_2 (plot PIP)]
depends: R_library('dscrutils') # can be installed via `devtools::install_github("stephenslab/dsc",subdir = "dscrutils", force = TRUE)`
output: f'{_input:n}.pdf'

task: trunk_workers = 1, trunk_size = job_size * 3, walltime = '2m', mem = '2G', cores = 1, tags = f'{step_name}_{_output:bn}'

R: expand = '${ }', stderr = f'{_output:n}.log', stdout = f'{_output:n}.stdout'
    s = dscrutils::read_dsc(${_input:r})
    # print something to log
    write(format(Sys.time(), "%a %b %d %X %Y"), stderr())
    library(susieR)
    is_susie = !is.null(s$elbo)
    if (is_susie) {
        write(paste('SuSiE objective:', susie_get_objective(s), "at L =", nrow(s$alpha)), stderr())
    } else {
        snp = s$snp
        snp = snp[match(s$var_names, snp$snp),]
        s$pip = snp$snp_prob
        class(s) = 'susie'
    }
    # make some plots
    pdf(${_output:r}, 8,9,pointsize=16)
    par(mar=c(2,4,0,0), omi=c(0,0,0,0), oma=c(0,0,0,0),mfrow=c(2,1))
    susie_plot(s,y='PIP',add_legend=T)
    susie_plot(s,y='z')
    dev.off()
    ordering = order(s$pip, decreasing=T)
    if (is.null(s$prior)) s$prior = rep(1/length(ordering),length(ordering))
    for (i in 1:5) {
        write(paste0("#", i, " pip-ranked SNP ", s$var_names[ordering[i]], " has z-score ", 
              round(s$z[ordering[i]],2), " pip ", round(s$pip[ordering[i]],4), " and prior ", round(s$prior[ordering[i]],4)), stderr())
    }
    ordering = order(abs(s$z), decreasing=T)
    for (i in 1:5) {
        write(paste0("#", i, " z-ranked SNP ", s$var_names[ordering[i]], " has z-score ", 
              round(s$z[ordering[i]],2), " pip ", round(s$pip[ordering[i]],4), " and prior ", round(s$prior[ordering[i]],4)), stderr())
    }
</code></pre></div></div>

<h2 id="example-usage">Example usage</h2>

<p>Commands below run the pipeline on a local computer.</p>

<p><strong>For UChicago midway users</strong>: to run these pipelines on RCC cluster, please take a look at <code class="highlighter-rouge">midway2.yml</code> file (found in this repository), modify it as you see fit, and replace <code class="highlighter-rouge">-q none -j 8</code> with:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>-c workflow/midway2.yml -q midway2 -J 40
</code></pre></div></div>

<h3 id="susie-hatb">SuSiE $\hat{b}$</h3>

<p>GWAS example run (with annotations):</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/summary_statistics_finemapping.ipynb susie_bhat \
        --loci /home/gaow/tmp/19-Dec-2018/SCZ/chunks.list \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics \
        --annotation atac-seq \
        --n 79845 \
        --susie-pve 0.005 --L 5 \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32msusie_bhat_1[0m: Run SuSiE bhat version
INFO: [32msusie_bhat_1[0m (index=0) is [32mignored[0m due to saved signature
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.rds[0m
INFO: Running [32msusie_bhat_2[0m: Summarize Fine-mapping results
INFO: [32msusie_bhat_2[0m (index=0) is [32mignored[0m due to saved signature
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.pdf[0m
INFO: Workflow susie_bhat (ID=897b7f9c97b1a5d0) is ignored with 2 ignored steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.pdf (7.6 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_25_2.png" alt="png" /></p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.log
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.log</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_B.L_5.prior_0p005.res_var_true.log (837 B):</div>
</div>

<div class="output output_html">
<div class="sos_hint">12 lines (5 displayed, see --limit)</div>
</div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Fri Dec 21 03:27:56 PM 2018
SuSiE objective: -113313.785692437 at L = 5
#1 pip-ranked SNP 9:84789089:T:C has z-score -5.77 pip 0.123 and prior 1e-04
#2 pip-ranked SNP 9:84700694:T:C has z-score -2.42 pip 0.1168 and prior 1e-04
#3 pip-ranked SNP 9:84725047:T:C has z-score -1.9 pip 0.1166 and prior 1e-04
</code></pre></div></div>

<p>eQTL example run (without annotations):</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/summary_statistics_finemapping.ipynb susie_bhat \
        --loci /home/gaow/tmp/19-Dec-2018/metasoft/one_loci.txt \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats \
        --n 800 \
        --susie-pve 0.1 --L 5 \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32msusie_bhat_1[0m: Run SuSiE bhat version
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.uniform.SuSiE_B.L_5.prior_0p1.res_var_false.rds[0m
INFO: Running [32msusie_bhat_2[0m: Summarize Fine-mapping results
null device 
          1 
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.uniform.SuSiE_B.L_5.prior_0p1.res_var_false.pdf[0m
INFO: Workflow susie_bhat (ID=10c68937a3df62e9) is executed successfully with 2 completed steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.uniform.SuSiE_B.L_5.prior_0p1.res_var_false.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.uniform.SuSiE_B.L_5.prior_0p1.res_var_false.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.uniform.SuSiE_B.L_5.prior_0p1.res_var_false.pdf (4.8 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_29_2.png" alt="png" /></p>

<h3 id="susie-z">SuSiE $z$</h3>

<p>GWAS example run (with annotations):</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sos run workflow/summary_statistics_finemapping.ipynb susie_z \
        --loci /home/gaow/tmp/19-Dec-2018/SCZ/chunks.list \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics \
        --annotation atac-seq \
        --L 5 \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32msusie_z_1[0m: Run SuSiE Z score version
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_Z.L_5.rds[0m
INFO: Running [32msusie_z_2[0m: Summarize Fine-mapping results
null device 
          1 
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_Z.L_5.pdf[0m
INFO: Workflow susie_z (ID=faeee78a5b518cd1) is executed successfully with 2 completed steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_Z.L_5.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_Z.L_5.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.SuSiE_Z.L_5.pdf (7.2 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_33_2.png" alt="png" /></p>

<h3 id="dap">DAP</h3>

<p>GWAS example run (with annotations):</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>DAP_EXE_PATH=~/GIT/github/mvarbvs/dsc/modules/linux
sos run workflow/summary_statistics_finemapping.ipynb dap \
        --loci /home/gaow/tmp/19-Dec-2018/SCZ/chunks.list \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics \
        --annotation atac-seq \
        --n 79845 \
        -b $DAP_EXE_PATH \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32mdap_1[0m: Run DAP-G
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.DAP.pkl[0m
INFO: Running [32mdap_2[0m: Summarize Fine-mapping results
null device 
          1 
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.DAP.pdf[0m
INFO: Workflow dap (ID=c895e1125141fae4) is executed successfully with 2 completed steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.DAP.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.DAP.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.atac-seq.DAP.pdf (7.0 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_37_2.png" alt="png" /></p>

<h3 id="caviar-1">CAVIAR</h3>

<p>GWAS example run:</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CAVIAR_EXE_PATH=~/GIT/github/mvarbvs/dsc/modules/linux
sos run workflow/summary_statistics_finemapping.ipynb caviar \
        --loci /home/gaow/tmp/19-Dec-2018/SCZ/chunks.list \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics \
        -b $CAVIAR_EXE_PATH --args "'-c 2'" \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32mcaviar_1[0m: Run CAVIAR
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.CAVIAR._c_2.rds[0m
INFO: Running [32mplot PIP[0m: Summarize Fine-mapping results
null device 
          1 
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.CAVIAR._c_2.pdf[0m
INFO: Workflow caviar (ID=23c609e929d3e8b1) is executed successfully with 2 completed steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.CAVIAR._c_2.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.CAVIAR._c_2.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.CAVIAR._c_2.pdf (7.1 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_41_2.png" alt="png" /></p>

<p>eQTL example run:</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>CAVIAR_EXE_PATH=~/GIT/github/mvarbvs/dsc/modules/linux
sos run workflow/summary_statistics_finemapping.ipynb caviar \
        --loci /home/gaow/tmp/19-Dec-2018/metasoft/one_loci.txt \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats \
        -b $CAVIAR_EXE_PATH --args "'-c 2'" \
        -q none -j 8
</code></pre></div></div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>INFO: Running [32mcaviar_1[0m: Run CAVIAR
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.CAVIAR._c_2.rds[0m
INFO: Running [32mplot PIP[0m: Summarize Fine-mapping results
null device 
          1 
INFO: output:   [32m/home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.CAVIAR._c_2.pdf[0m
INFO: Workflow caviar (ID=157b4cf09027e3f5) is executed successfully with 2 completed steps.

</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.CAVIAR._c_2.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.CAVIAR._c_2.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.CAVIAR._c_2.pdf (4.7 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_44_2.png" alt="png" /></p>

<h3 id="finemap-1">FINEMAP</h3>

<p>GWAS example run:</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>FINEMAP_EXE_PATH=~/GIT/github/mvarbvs/dsc/modules/linux
sos run workflow/summary_statistics_finemapping.ipynb finemap \
        --loci /home/gaow/tmp/19-Dec-2018/SCZ/chunks.list \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics \
        -b $FINEMAP_EXE_PATH --n 500 \
        -q none -j 8
</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.FINEMAP.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.FINEMAP.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/SCZ/Summary_statistics/chr9_84630941_84813641/chr9_84630941_84813641.FINEMAP.pdf (7.0 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_48_2.png" alt="png" /></p>

<p>eQTL example run:</p>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>FINEMAP_EXE_PATH=~/GIT/github/mvarbvs/dsc/modules/linux
sos run workflow/summary_statistics_finemapping.ipynb finemap \
        --loci /home/gaow/tmp/19-Dec-2018/metasoft/one_loci.txt \
        --ss-data-prefix /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats \
        -b $FINEMAP_EXE_PATH --n 500 \
        -q none -j 8
</code></pre></div></div>

<div class="input_area highlighter-rouge"><div class="highlight"><pre class="highlight"><code>%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.FINEMAP.pdf -s png --dpi 50
</code></pre></div></div>

<div class="output output_html">
<div class="sos_hint">%preview /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.FINEMAP.pdf</div>
</div>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.FINEMAP.pdf (4.6 KiB):</div>
</div>

<p class="output output_png"><img src="../images/finemapping/summary_statistics_finemapping_51_2.png" alt="png" /></p>

<div class="output output_html">
<div class="sos_hint">&gt; /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.FINEMAP.pdf:</div>
</div>

<div class="output output_stream highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bash: /home/gaow/tmp/19-Dec-2018/metasoft/gtex_metasoft_summary_stats/ENSG00000181404.15/ENSG00000181404.15.FINEMAP.pdf: Permission denied

</code></pre></div></div>

              <nav class="c-page__nav">
  
    
    <a id="js-page__nav__prev" class="c-page__nav__prev" href="/jupyterbook-sos-demo/finemapping/summary_statistics_wrangler">
      〈 <span class="u-margin-right-tiny"></span> Terminology
    </a>
  

  
    
    <a id="js-page__nav__next" class="c-page__nav__next" href="/jupyterbook-sos-demo/finemapping/finemapping_results_wrangler">
      Running DSC <span class="u-margin-right-tiny"></span> 〉
    </a>
  
</nav>

            </div>
          </div>
        </div>
      </main>
    </div>

  </body>
</html>
